docs/keps/20200619-kubefed-pull-reconciliation.md

```diff
@@ -0,0 +1,362 @@
+This differs from the current architecture where the control-plane has write access
+to the kubefed clusters.
+
+Consequently the `control-plane` and `kubefed daemons` need to communicate between themselves.
+A bi-directional communication is required, consequently a secure trust communication should
+be established between the control-plane and daemons. In the future, a good solution would be to follow the [SPIFEE Trust domain federation](https://docs.google.com/document/d/1OC9nI2W04oghhbEDJpKdIUIw-G23YzWeHZxwGLIkB8k/edit) approach.
+To federate identity and trust, you must exchange the trust bundles between the `control-plane`
+and `kubefed daemon` of each cluster.
+
+
+### User Stories
+
+#### Story 1
+
+Users want to avoid giving cluster admin access to something like federation,
+some admins have expressed distaste at that requirement.
+If the management cluster is compromised that would give access to the kubefed clusters.
+
+With this new proposal, the management cluster only has read access to the kubefed clusters.
+
+#### Story 2
+
+Kubefed uses push reconciliation, all the workload relies on the control plane running
+on the management cluster. Due to the existing design decisions, kubefed represents
+a bottleneck when having to manage thousands of resources across many clusters.
+
+With this new proposal, Kubefed would use a pull-based reconciliation mechanism
+and most of the workload is shared between the kubefed agents and the control-plane.
+
+#### Story 3
+
+Customers might have an infrastructure setup where the kubefed clusters are behind
+some NAT gateways, and consequently the control-plane cannot reach the kubefed clusters.
+
+With this new approach, the kubefed daemon can register with the control plane and sets up
+a bi-directional tunnel.
+
+## Alternatives
+
+### Centralized Resource Management -- GitOps
+
+In order to avoid the `kubefed daemon --> control-plane` communication, I thought
+about using a centralized component where the federated resources could be created
+to be consumed by all the kubefed clusters.
+
+This centralized component represents a hybrid approach of using the federation api to define configuration, and then generating a canonical form into a git repository for gitOps.
+This alternative solution might represent a good solution for some use cases.
+
+#### Analysis
+
+As an administrator, the admins prefer to have the RBAC rules in the cluster,
+and not outsourced to a git repository with rules about who gets to push what where.
+
+With this alternative, the communication flow goes against the gitOps repository or
+centralized datastore where the desired state is stored.
+
+This alternative adds a new third party in the game which might increase the complexity of the whole architecture.
+
+In terms of security, the system should rely on what each cluster is allowed to view,
+likewise it is hard to understand how `overwrites` would work in a per cluster level.
```

Comment:
Why would `overrides` be any different from now. Each cluster reads the federated resource together with a placement and the associated override and computes the finalised resource for itself.

Reply:
Overrides right now are hard to manage in cases where every cluster has a different override (we have use cases like that), but this is not in the scope of this KEP IMO.